composer/algorithms/alibi/attention_surgery_functions/utils.py

```diff
@@ -0,0 +1,171 @@
+        if not isinstance(old_weight, torch.nn.Parameter):
+            raise TypeError(
+                f"Module {module._get_name()}, position embedding {position_embedding_attribute}, "
+                f"'weight' attribute must be of type torch.nn.Module"
+            )
+        new_weight = torch.nn.Parameter(
+            torch.zeros((max_sequence_length, old_weight.shape[1]),
+                        dtype=old_weight.dtype,
+                        layout=old_weight.layout,
+                        device=old_weight.device))
+        new_weight.requires_grad = False
+        setattr(pos_embedding_module, 'weight', new_weight)
+
+        log.info(f' Position embedding expanded to sequence length {max_sequence_length}, zeroed, and frozen')
+
+    except AttributeError:
+        # TODO: This should perhaps result in an exception rather than a warning
+        log.error(f'Unable to zero and freeze position embeddings. Module '
+                  f'{module} may lack attribute {position_embedding_attribute}, or position '
+                  f"embeddings may lack attribute 'weight'.")
+
+
+def register_alibi(module: torch.nn.Module, n_heads: int, max_token_length: int, causal: bool) -> torch.nn.Module:
+    """Adds ALiBi's linear attention biases as a buffer to the module."""
+    if causal:  # e.g., for GPT
+        # Modified from https://github.com/ofirpress/attention_with_linear_biases/blob/5b327adc6d131e28b40ba58906b30bb469483519/fairseq/models/transformer.py#L742
+        slopes = torch.Tensor(_get_alibi_head_slopes(n_heads))
+        # In the next line, the part after the * is what constructs the diagonal matrix
+        # (right matrix in Figure 3 in the paper).
+        # If you run it you'll see that it doesn't exactly print out the same matrix as we
+        # have in Figure 3, but one where all rows are identical.
+        # This works because the softmax operation is invariant to translation, and our bias
+        # functions are always linear.
+        alibi = slopes.unsqueeze(1).unsqueeze(1) * \
+            torch.arange(max_token_length). \
+            unsqueeze(0).unsqueeze(0).expand(n_heads, -1, -1)
+
+    else:  # e.g., for BERT
+        # Following https://github.com/ofirpress/attention_with_linear_biases/issues/5 (Implementation 1)
+        # In the causal case, you can exploit the fact that softmax is invariant to a uniform translation
+        # of the logits, which makes the math work out *after* applying causal masking. If no causal masking
+        # will be applied, it is necessary to construct the diagonal mask.
+        context_position = torch.arange(max_token_length)[:, None]
+        memory_position = torch.arange(max_token_length)[None, :]
+        relative_position = torch.abs(memory_position - context_position)
+        # [n_heads, max_token_length, max_token_length]
+        relative_position = relative_position.unsqueeze(0).expand(n_heads, -1, -1)
```

Comment:
This is the kind of thing I would write a test for, but I'm also not very good at mentally reshaping tensors.

Reply:
Added some asserts to confirm sizing. Any sizing mistakes should also cause unit testing to fail.